subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy

```diff
@@ -174,6 +174,57 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
         outputDirectory.list() == [input.name]
     }
 
+    @Issue('https://github.com/gradle/gradle/issues/9904')
+    def "task with broken symlink in inputs are valid"() {
+        def inputDirectoryWithBrokenLink = file('inputDirectoryWithBrokenLink').createDir()
+        def inputFileTarget = file("brokenInputFileTarget")
+        def inputFile = inputDirectoryWithBrokenLink.file('brokenInputFile').createLink(inputFileTarget)
+        def output = file("output.txt").createFile()
+
+        // Validation fails with broken links assigned directly to `@InputDirectory` or `@InputFile`
+        buildFile << """
+            class CustomTask extends DefaultTask {                
+                @InputFiles FileCollection brokenInputFile
+
+                @InputDirectory File inputDirectoryWithBrokenLink
+
+                @OutputFile File output
+
+                @TaskAction execute() {
+                    output.text = "\${brokenInputFile.files*.name} \${inputDirectoryWithBrokenLink.name} \${inputDirectoryWithBrokenLink.list()}"
+                }
+            }
+            task inputBrokenLinkNameCollector(type: CustomTask) {
+                inputDirectoryWithBrokenLink = file '${inputDirectoryWithBrokenLink}'
+                brokenInputFile = files '${inputFile}'
+                output = file '${output}'
+            }
+        """
+        assert inputDirectoryWithBrokenLink.exists() && !inputFile.exists() && output.text.empty
+
+        when:
+        run 'inputBrokenLinkNameCollector', '--info'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+        output.text == "[brokenInputFile] inputDirectoryWithBrokenLink [brokenInputFile]"
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
+
+        when:
+        inputFileTarget.createFile()
```

Comment:
We currently don't know if the change is caused by something detected in the file collection or in the input directory. We should probably change both things separately or have separate tests for each of the cases. A `where` clause would probably work good here.

Reply:
Valid concerns, separated via `where`